src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -2499,3 +2500,82 @@ bool ShenandoahHeap::requires_barriers(stackChunkOop obj) const {
 
   return false;
 }
+
+HeapWord* ShenandoahHeap::allocate_loaded_archive_space(size_t size) {
+  // CDS wants a continuous memory range to load a bunch of objects.
+  // This effectively bypasses normal allocation paths, and requires
+  // a bit of massaging to unbreak GC invariants.
+
+  ShenandoahAllocRequest req = ShenandoahAllocRequest::for_shared(size);
+
+  // Easy case: a single regular region, no further adjustments needed.
+  if (size < ShenandoahHeapRegion::humongous_threshold_words()) {
+    return allocate_memory(req);
+  }
+
+  // Hard case: the requested size would cause a humongous allocation.
+  // We need to make sure it looks like regular allocation to the rest of GC.
+
+  // CDS code would guarantee no objects straggle multiple regions, as long as
+  // regions are as large as MIN_GC_REGION_ALIGNMENT. It is impractical at this
+  // point to deal with case when Shenandoah runs with smaller regions.
+  // TODO: This check can be dropped once MIN_GC_REGION_ALIGNMENT agrees more with Shenandoah.
+  if (ShenandoahHeapRegion::region_size_bytes() < ArchiveHeapWriterConsts::MIN_GC_REGION_ALIGNMENT) {
+    return nullptr;
+  }
+
+  HeapWord* mem = allocate_memory(req);
+  size_t start_idx = heap_region_index_containing(mem);
+  size_t num_regions = ShenandoahHeapRegion::required_regions(size * HeapWordSize);
+
+  // Flip humongous -> regular.
+  {
+    ShenandoahHeapLocker locker(lock(), false);
+    for (size_t c = start_idx; c < start_idx + num_regions; c++) {
+      get_region(c)->make_regular_bypass();
+    }
+  }
+
+  // If the trailing region is not full, we need to adjust its top.
+  size_t tail = (size % ShenandoahHeapRegion::region_size_words());
```

Comment:
Not sure we need this. The free set adjusts top for all humongous regions in the allocation (including the tail): https://github.com/openjdk/jdk/blob/master/src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp#L885. What's more, I believe `new_top` is only used during a full GC to sort out `top` once compaction is complete.

Reply:
Right, d'oh. Not sure what I was thinking there. I'll drop the block, since the whole thing apparently works without adjusting the top.